server/api_archive.go

```diff
@@ -0,0 +1,197 @@
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+// The contents of a README file which will be inserted into the archive.
+const ReadmeContents = `This archive contains uninterpreted server files.
+These files all contain publicly available information
+and additional work is needed to stand up a new server
+using them.
+`
+
+// ArchiveHandler provides the POST /archive api endpoint. It returns
+// uninterpreted file data intended to be used by an external service
+// to produce an archive.
+func (gcas *GCAServer) ArchiveHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		http.Error(w, "Only GET method is allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	if r.ContentLength != 0 {
+		http.Error(w, "Request should not have a body", http.StatusBadRequest)
+		return
+	}
+
+	if !gcas.ApiArchiveRateLimiter.Allow() {
+		http.Error(w, fmt.Sprintf("Too many requests, this server allows %v every %v", apiArchiveLimit, apiArchiveRate), http.StatusTooManyRequests)
+		return
+	}
+
+	buffer := new(bytes.Buffer)
+	znw := zip.NewWriter(buffer)
+
+	// File ordering matters here. We are not locking the files, and relying on the file writes being
+	// append only while happening in a single write. We must archive the files in reverse order from
+	// their dependency order.
+
+	for _, name := range PublicFiles {
+		err := gcas.addFile(name, znw)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Error zipping file %v: %v", name, err), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	// Add the pseudo file server.pubkey last, as the other files are signed by it.
+	if err := gcas.addPubKeyFile("server.pubkey", znw); err != nil {
+		http.Error(w, fmt.Sprintf("Error zipping server.pubkey: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	// Add a README
+	if err := gcas.addReadmeFile(znw); err != nil {
+		http.Error(w, fmt.Sprintf("Error adding README: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	if err := znw.Close(); err != nil {
+		if _, err := w.Write(buffer.Bytes()); err != nil {
+			http.Error(w, "Failed to close zip writer", http.StatusInternalServerError)
+			return
+		}
+	}
+
+	w.Header().Set("Content-Type", "application/zip")
+
+	if _, err := w.Write(buffer.Bytes()); err != nil {
+		http.Error(w, "Failed to write archive response", http.StatusInternalServerError)
+	}
+}
+
+// Add a file to a zip archive, copying all the data in the file.
+func (gcas *GCAServer) addFile(name string, zipw *zip.Writer) error {
+	path := filepath.Join(gcas.baseDir, name)
+
+	file, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return err
+	}
+
+	header, err := zip.FileInfoHeader(fileInfo)
+	if err != nil {
+		return err
+	}
+
+	header.Method = zip.Deflate
+
+	writer, err := zipw.CreateHeader(header)
+	if err != nil {
+		return err
+	}
+
+	// File contains only public data so copy it all.
```

Comment:
This comment says the file contains only public data, but there's actually nothing in the implementation of `addFile` that enforces that. So I'd probably just remove this comment.